api/src/main/java/org/openmrs/module/debezium/service/DebeziumEventQueueServiceImpl.java

```diff
@@ -0,0 +1,163 @@
+	
+	private static final Logger logger = LoggerFactory.getLogger(DebeziumEventQueueServiceImpl.class);
+	
+	DebeziumEventQueueDAO eventQueueDAO;
+	
+	DebeziumEventQueueOffsetDAO offsetDAO;
+	
+	@Override
+	public Set<DebeziumEventQueue> getEventsByApplicationName(String applicationName) {
+		
+		Set<DebeziumEventQueue> eventQueues = new HashSet<>();
+		List<String> parameterizedTables = getParameterizedTables(applicationName);
+		DebeziumEventQueueOffset offset = offsetDAO.getOffsetByApplicationName(applicationName);
+		int fetchSize = Integer.parseInt(getFetchSize());
+		
+		// Request new data without commit of the preview request
+		if (offset != null && offset.getLastRead() != null) {
+			return this.processEventWithoutCommit(offset, parameterizedTables);
+		} else {
+			// Request of new data without offset respecting the fetch size limit
+			while (eventQueues.size() < fetchSize) {
+				List<DebeziumEventQueue> eventQueue;
+				if (offset != null && offset.isCreated() && offset.getLastRead() != null) {
+					eventQueue = eventQueueDAO.getEventsByApplicationNameRecursive(offset.getLastRead(), fetchSize);
+					
+				} else {
+					eventQueue = eventQueueDAO.getEventsByApplicationName(offset, fetchSize);
+				}
+				
+				if (eventQueue == null || eventQueue.isEmpty()) {
+					break;
+				}
+				
+				for (DebeziumEventQueue debeziumEventQueue : eventQueue) {
+					if (parameterizedTables.contains(debeziumEventQueue.getTableName())) {
+						eventQueues.add(debeziumEventQueue);
+						if (eventQueues.size() >= fetchSize) {
+							break;
+						}
+					}
+				}
+				//update offset or create new one
+				if (!eventQueues.isEmpty()) {
+					if (offset != null) {
+						offset.setLastRead(eventQueue.get(eventQueue.size() - 1).getId());
+					} else {
+						offset = new DebeziumEventQueueOffset();
+						offset.setFirstRead(eventQueue.get(0).getId());
+						offset.setLastRead(eventQueue.get(eventQueue.size() - 1).getId());
+						offset.setApplicationName(applicationName);
+						offset.setActive(Boolean.TRUE);
+						offset.setCreatedAt(new Date());
+					}
+				}
+			}
+		}
+		
+		if (!eventQueues.isEmpty()) {
+			if (offset.isCreated()) {
+				offsetDAO.updateOffset(offset);
+			} else {
+				offsetDAO.saveOffset(offset);
+			}
+		}
+		
+		return eventQueues;
+	}
+	
+	private Set<DebeziumEventQueue> processEventWithoutCommit(DebeziumEventQueueOffset offset,
+	        List<String> parameterizedTables) {
+		Set<DebeziumEventQueue> previousEventQueues = new HashSet<>();
+		List<DebeziumEventQueue> alreadyProcessedEvents = eventQueueDAO.getEventsByApplicationName(offset, null);
+		
+		for (DebeziumEventQueue debeziumEventQueue : alreadyProcessedEvents) {
+			if (parameterizedTables.contains(debeziumEventQueue.getTableName())) {
+				previousEventQueues.add(debeziumEventQueue);
+			}
+		}
+		return previousEventQueues;
+	}
+	
+	@Override
+	public void commitEventQueue(String applicationName) {
```

Comment:
The commit has nothing to do with the queue, it should be placed on the offset class